### ISSUE TITLE
Hotkeys Configfile

### ISSUE THREAD

**Issue by Krarilotus** [Sh0wdown/UnofficialCrusaderPatch#432]
This is probably a big change regarding online play and Multiplayer in general.
As we might see balance changes made possible by a config at some point, which is also heck of alot of work, could it be possible to also have a config for hotkeys?

What i mean: have a 'hotkeys' tab and select your own hotkeyconfig, which you could share with others too. There you could redefine existing hotkeys like for example make a grid layout instead of spreading everything over the keyboard (M for market, A for Armory..., X and C for turning the screen... So instead of N, B and I to recruit troops, you would have Q,W,E...)

If you need, i will post all existing hotkeys here in this issue.

**Comment by LeSpec** [comment]
In multiplayer it would be nice to have a shortcut for the assassin flag, since they move faster when going to the flag.

**Comment by GRhin** [comment]
You mean like ctrl+n followed by the number 4?

**Comment by LeSpec** [comment]
Yes, exactly

**Comment by GRhin** [comment]
Have you tried that combination? Or is it shift? I keep getting shift and
ctrl mixed up for that hotkey.

**Comment by LeSpec** [comment]
Yes, ctrl+n
What i mean is pressing one key instead of that combination, like you also select your groups by pressing only one key.
Replacing ctrl+n with n makes also sense practicly. No one realy wants to switch the screen to the building, you only want to use the menu.

**Comment by GRhin** [comment]
Fair enough. Hotkeys are a long way off I think. None of the devs have touched any of the code around handling user input,  so it will take a fair bit to familiarize themselves with it before they know what can be done there.

**Comment by LeSpec** [comment]
So i knew that wasd keys and quick saves were already patched which kinda encouraged me to look into it.
I replaced tab functioning with selecting the assassin flag and switched the ctrl functioning of h, m, b, i, t, n, g and a.
Patched it by hand, haven't looked into Sh0wdown's tools yet.

position of control switches:

**sub_4B2AE0**
![Bildschirmfoto 2020-12-11 um 08 54 18](https://user-images.githubusercontent.com/62205974/101880133-b43b9600-3b92-11eb-8bf5-de8f8ab17e12.png)

for the assassin flag:
0x1FE7AEC = 0x16B (value of selected item)

**Comment by GRhin** [comment]
From what I remember and understand from when the devs were discussing it, raplacing existing hotkeys is simple, which is what the wasd hotkeys are. The problem is when adding new hotkeys, and toplevel hotkey to assassin rally point is new. 
I dont think it should be done tbh, as i think assassins should move at same speed always, which is another fix, which effects balance.
If you are going to do for assassins you should do for all units, and there isnt enough keys on the keyboard for all those.
Other games have a similar sorting system, press key 1 to get to this category of commands, then click key 2 to do specific action, and this I think is the best behaviour for rally pointing, which is already implemented via shift+n followed by 4

**Comment by J-T-de** [comment]
@LeSpec really cool findings, I am not aware if any of the developers has tried to do the changing hotkey things lately. I am only aware that LordHansCarpon looked into how to potentially introduce new hotkeys, but iirc this lead to nothing.

However, I think alone a "remapping" of existing hotkeys is really useful (the laptop folks and https://github.com/Sh0wdown/UnofficialCrusaderPatch/issues/179 ...)

> for the assassin flag:
0x1FE7AEC = 0x16B (value of selected item)

This is HUGE: This literally allows selecting everything. Even a new keep... this would actually be useful for a lot of hotkeys similar to the shcplayer. If you are interested in a faster communication, just pass by the UCP discord ;)
